delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/resources/ResourceValidator.scala

```diff
@@ -0,0 +1,75 @@
+package ch.epfl.bluebrain.nexus.delta.sdk.resources
+
+import ch.epfl.bluebrain.nexus.delta.rdf.IriOrBNode.Iri
+import ch.epfl.bluebrain.nexus.delta.rdf.Vocabulary.{contexts, schemas}
+import ch.epfl.bluebrain.nexus.delta.rdf.graph.Graph
+import ch.epfl.bluebrain.nexus.delta.rdf.jsonld.ExpandedJsonLd
+import ch.epfl.bluebrain.nexus.delta.rdf.jsonld.api.JsonLdApi
+import ch.epfl.bluebrain.nexus.delta.rdf.shacl.{ShaclEngine, ValidationReport}
+import ch.epfl.bluebrain.nexus.delta.sdk.identities.model.Caller
+import ch.epfl.bluebrain.nexus.delta.sdk.model.ResourceF
+import ch.epfl.bluebrain.nexus.delta.sdk.resolvers.ResolverResolution.ResourceResolution
+import ch.epfl.bluebrain.nexus.delta.sdk.resources.model.ResourceRejection
+import ch.epfl.bluebrain.nexus.delta.sdk.resources.model.ResourceRejection.{InvalidJsonLdFormat, InvalidResource, InvalidSchemaRejection, ReservedResourceId, ResourceShaclEngineRejection, SchemaIsDeprecated}
+import ch.epfl.bluebrain.nexus.delta.sdk.schemas.model.Schema
+import ch.epfl.bluebrain.nexus.delta.sourcing.model.ResourceRef.Latest
+import ch.epfl.bluebrain.nexus.delta.sourcing.model.{ProjectRef, ResourceRef}
+import monix.bio.IO
+
+final class ResourceValidator(resourceResolution: ResourceResolution[Schema])(implicit
+    jsonLdApi: JsonLdApi
+) {
+  private def toGraph(id: Iri, expanded: ExpandedJsonLd): IO[ResourceRejection, Graph] =
+    IO.fromEither(expanded.toGraph).mapError(err => InvalidJsonLdFormat(Some(id), err))
+
+  def validationReport(
+      projectRef: ProjectRef,
+      schemaRef: ResourceRef,
+      caller: Caller,
+      resourceId: Iri,
+      expanded: ExpandedJsonLd
+  ): IO[ResourceRejection, (Option[ValidationReport], ResourceRef.Revision, ProjectRef)] =
```

Comment:
A case class for the return method would be better

Reply:
I am planning to change this into two different calls, but it's possible yes

Reply:
I was thinking this: 
`def validationReport(
                        projectRef: ProjectRef,
                        schemaRef: ResourceRef,
                        caller: Caller,
                        resourceId: Iri,
                        expanded: ExpandedJsonLd
                      ): IO[ResourceRejection, Option[ValidationReport]]`

and

`def validate(
                        projectRef: ProjectRef,
                        schemaRef: ResourceRef,
                        caller: Caller,
                        resourceId: Iri,
                        expanded: ExpandedJsonLd
                      ): IO[ResourceRejection, (ResourceRef.Revision, ProjectRef)]`

Reply:
I think it would be better to have something like this as a return type:
```
sealed trait ValidationResult
case object NoValidation extends ValidationResult
case object Validated(schema, validationReport)
```
It is valuable in the response to have a proper message if there was no validation too.
And if there is one besides having the validation report to have information about the schema (remember that the schema could live in another project)

NB: feel free to change the names,